### DISCUSSION TITLE
Is it possible to use postMessage from python code executed in webworker?

### DISCUSSION THREAD

**Discussion by nioate** [pyodide/pyodide#2346] [Q&A] [answered]
I would like to send messages (e.g. image data) from a loop inside python code to the "main" javascript thread and also receive messages (e.g. mouse click coordinates) which should then be processed in python. Thus, the python program should run permanently but be able to interact with the DOM via messages.
So far it seems as if one can only use postMessage from the webworker javascript code **after** runPython(...) finished, but not from within python.
Or is it somehow possible, which would probably require getting access to the webworker's self.postMessage function inside python?

**Comment by hoodmane** [comment, marked as answer]
Yeah it's possible:
```py
from js import postMessage as _post_message, addEventListener as _add_event_listener
from pyodide import to_js, create_proxy

def add_event_listener(event : str, listener : Callable[[Any], None]):
	_add_event_listener(event, create_proxy(listener))

def post_message(**kwargs):
	_post_message(to_js(kwargs))

post_message(cmd="test", data={ "name" : "nioate", "info" : "hi there!" })

def on_message(event):
	print(event.data.message)

add_event_listener("message", on_message)
```

**Reply by hoodmane** [reply]
Note that message events will only be processed when no other Python code is running though.
